### PR TITLE
Added MongoStorage for FSM

### PR DESCRIPTION
**Pull request by Rishat-F**
# Description

1. Added MongoStorage for FSM
2. Added tests for it
3. Docs updated
4. Some code was restructured

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (typos, code examples or any documentation update)

# How Has This Been Tested?

1. New tests added
2. Added few additional checks in existing tests
3. Code checked with black, isort, mypy

**Test Configuration**:
* Operating System: Ubuntu 20.04 LTS
* Python version: 3.10

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: aiogram/fsm/storage/base.py
Comment (JrooTJunior) [inline aiogram/fsm/storage/base.py:76]:
```suggestion
                "Default key builder is not configured to use key destiny other than the default.\n"
```

File: aiogram/fsm/storage/mongo.py
Comment (JrooTJunior) [inline aiogram/fsm/storage/mongo.py:72]: What about using single object key instead of creating multiple documents for the same user?

Just use:
```json5
{
    _id: "fsm:bot_id:user_id:chat_id:destiny", 
    state: state,
    data: {...}
}
```

instead of 
```json5
{
    _id: "fsm:bot_id:user_id:chat_id:destiny:state", 
    state: state,
}
```
and
```json5
{
    _id: "fsm:bot_id:user_id:chat_id:destiny:data",
    data: {...}
}
```

File: aiogram/fsm/storage/mongo.py
Comment (JrooTJunior) [inline aiogram/fsm/storage/mongo.py:107]: Why type ignored?

Comment (JrooTJunior) [inline aiogram/fsm/storage/mongo.py:107]: This statement can be simplified
```suggestion
        if not document:
            return {}
        return document
```

File: aiogram/fsm/storage/mongo.py
Comment (JrooTJunior) [inline aiogram/fsm/storage/mongo.py:120]: Why ignore type?

**Review by JrooTJunior — CHANGES_REQUESTED** [review]

File: aiogram/fsm/storage/mongo.py
Comment (Rishat-F) [reply aiogram/fsm/storage/mongo.py:107]: Mypy check throws error:
error: Incompatible return value type (got "Mapping[str, Any]", expected "Dict[str, Any]")

File: aiogram/fsm/storage/mongo.py
Comment (Rishat-F) [reply aiogram/fsm/storage/mongo.py:72]: Sounds reasonable. I will rewrite it within a couple of days. And we will see how it will look like.

Comment (Rishat-F) [reply aiogram/fsm/storage/mongo.py:72]: I have thought about this proposal more carefully and came to the conclusion that it is better not to combine state and data into one collection. Due to the following reasons:
1. Most likely we don’t want to store documents with empty "state" key and empty "data" key at the same time (with only "_id" key), because it would be a waste of memory. If so, then there would be more code, more logic, more queries to define if document become empty and remove it.
2. It seems like there no way to update data with one query if it is stored as nested object. So there also needed more code and more queries to do it.
So the main advantage of using two separated collections for state and data over using one union collection is more simple code and fewer queries to MongoDB.

File: aiogram/fsm/storage/base.py
Comment (Olegt0rr) [inline aiogram/fsm/storage/base.py:28]: Redundant comment, `@abstractmethod` already meant this.
Use imperative method description only.

File: aiogram/fsm/storage/base.py
Comment (Olegt0rr) [inline aiogram/fsm/storage/base.py:34]: Redundant `pass`. This string may be removed

File: aiogram/fsm/storage/base.py
Comment (Olegt0rr) [inline aiogram/fsm/storage/base.py:39]: Docstring should started without linebreak

```suggestion
    """Simple key builder with default prefix.
```

File: aiogram/fsm/storage/base.py
Comment (Olegt0rr) [inline aiogram/fsm/storage/base.py:76]: Please, avoid direct passing string to an error to avoid traceback output duplications. Assign string to variable first.

File: aiogram/fsm/storage/mongo.py
Comment (Olegt0rr) [inline aiogram/fsm/storage/mongo.py:47]: Don't create mutable objects in default args

File: aiogram/fsm/storage/mongo.py
Comment (Olegt0rr) [reply aiogram/fsm/storage/mongo.py:72]: 2. https://www.mongodb.com/community/forums/t/how-to-update-a-nested-document-use-operator-in-mongodb/170700

File: aiogram/fsm/storage/base.py
Comment (Olegt0rr) [reply aiogram/fsm/storage/base.py:39]: Ah, other code uses the same format. Skipped.

File: aiogram/fsm/storage/base.py
Comment (Rishat-F) [reply aiogram/fsm/storage/base.py:34]: `pass` is used within `@abstractmethod` in current codebase in many places. Therefore, it's better to leave it, so that everything is uniform.
If it should be removed from every module in codebase, seems like it's needed separate Pull Request with refactoring.

File: aiogram/fsm/storage/mongo.py
Comment (Rishat-F) [reply aiogram/fsm/storage/mongo.py:74]: 2. Finally find how to do that - https://www.mongodb.com/community/forums/t/upsert-nested-object/204990

Comment (Rishat-F) [reply aiogram/fsm/storage/mongo.py:74]: 1. @Olegt0rr, I meant the following difficulty for the case if we store state and data together.
State was created:
`{
    _id: "fsm:bot_id:user_id:chat_id:destiny", 
    state: state
}`
Then data was created:
`{
    _id: "fsm:bot_id:user_id:chat_id:destiny", 
    state: state,
    data: {...}
}`
When state (data) would be removed we need to check if data (state) is presented in document. If data (state) not presented in document we should remove whole document, because no reason to store document with only "_id" key, like this:
`{
    _id: "fsm:bot_id:user_id:chat_id:destiny"
}`
I thought that the implementation of this behavior requires more code and additional logic.
But now i see, that maybe it would be not so complicated as I thought. Okay, I'll give it a try.

File: aiogram/fsm/storage/mongo.py
Comment (Olegt0rr) [inline aiogram/fsm/storage/mongo.py:38]: `self._db_name` - seems redundant

File: aiogram/fsm/storage/mongo.py
Comment (Olegt0rr) [inline aiogram/fsm/storage/mongo.py:42]: Get single database instance via `self._client[db_name]` first, then reuse it for getting collection objects.
It will make code easier to read

File: aiogram/fsm/storage/mongo.py
Comment (Rishat-F) [reply aiogram/fsm/storage/mongo.py:38]: It is used in tests teardown

File: aiogram/fsm/storage/mongo.py
Comment (Rishat-F) [reply aiogram/fsm/storage/mongo.py]: Done

**Review by JrooTJunior — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml
Comment (Rishat-F) [reply .github/workflows/tests.yml]: What is the difference between `tests` and `pypy-tests` sections in workflows/tests.yml? Seems like `pypy-tests `section does not needed, because tests already runs in `tests` section.

Comment (JrooTJunior) [reply .github/workflows/tests.yml]: Separated because additional conditions in the job makes it harder to read and maintain, separated workflows easiest way to manage it.

**Comment by Rishat-F** [comment]
@JrooTJunior, what is the status of PR? Do you wait something from me to do?

**Comment by JrooTJunior** [comment]
> Do you wait something from me to do?

Waiting for free time to review and test it before merge.

**Comment by Vadim-Khristenko** [comment]
Hi all,
@Rishat-F Can you merge Aiogram 3.5.0, which adds full support for Bot API 7.2 with your PR? It would be cool because your PR allows me to use MongoDB in FSM and not have to resort to installing Redis or anything else. But unfortunately, in your PR you only have Bot API 7.1, and Bot API 7.2 and Aiogram 3.5.0 have already been released.

@JrooTJunior If possible, can you tell me the approximate date when the merger will take place, if at all. It would just be nice to see this, for example, in 3.5.1 or just in the near future.

**Comment by Rishat-F** [comment]
@Vadim-Khristenko I'll try it in a couple of weeks.

**Comment by Rishat-F** [comment]
@Vadim-Khristenko aiogram 3.5.0 merged in this PR

**Comment by Rishat-F** [comment]
@JrooTJunior could we complete this PR?

**Review by JrooTJunior — APPROVED** [review]

**Pull request merged**
